source/darwin/secure_transport_tls_channel_handler.c

```diff
@@ -556,22 +594,29 @@ static int s_handle_shutdown(
     bool abort_immediately) {
     struct secure_transport_handler *secure_transport_handler = handler->impl;
 
-    if (dir == AWS_CHANNEL_DIR_WRITE) {
-        if (!abort_immediately && error_code != AWS_IO_SOCKET_CLOSED) {
-            AWS_LOGF_TRACE(AWS_LS_IO_TLS, "id=%p: shutting down write direction.", (void *)handler);
-            SSLClose(secure_transport_handler->ctx);
-        }
-    } else {
+    if (dir == AWS_CHANNEL_DIR_READ) {
         AWS_LOGF_DEBUG(
-            AWS_LS_IO_TLS,
-            "id=%p: shutting down read direction with error %d. Flushing queues.",
-            (void *)handler,
-            error_code);
+            AWS_LS_IO_TLS, "id=%p: shutting down read direction with error %d.", (void *)handler, error_code);
+        if (!abort_immediately && secure_transport_handler->negotiation_finished &&
+            !aws_linked_list_empty(&secure_transport_handler->input_queue) && slot->adj_right) {
+            s_initialize_read_delay_shutdown(handler, slot, error_code);
+            /* Early out, not complete the shutdown process for the handler until the handler processes the pending
+             * data. */
+            return AWS_OP_SUCCESS;
+        }
+        /* Flushing queues */
         while (!aws_linked_list_empty(&secure_transport_handler->input_queue)) {
```

Comment:
There's a possible memory leak right now:
- if we're doing delayed shutdown
- and we end up calling `aws_channel_slot_on_handler_shutdown_complete()` with some data remaining in `input_queue` (maybe there was an error from `SSLRead()`, or maybe there's some garbage data leftover after the graceful close frame)

Anway, maybe move this back where it was, in the common code at the end of the function, so it gets hit on immediate-read-shutdown OR when we swing back around in the write direction.